src/emitters/spot.cpp

```diff
@@ -0,0 +1,188 @@
+
+        return beam_and_cutoff_res;
+    }
+
+    std::pair<Ray3f, Spectrum> sample_ray(Float time, Float wavelength_sample,
+                                          const Point2f &spatial_sample,
+                                          const Point2f & /*dir_sample*/,
+                                          Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleRay, active);
+
+        // 1. Sample directional component
+        const Transform4f &trafo = m_world_transform->eval(time, active);
+        Vector3f local_dir = warp::square_to_uniform_cone(spatial_sample, (Float)m_cos_cutoff_angle);
+        Float pdf_dir = warp::square_to_uniform_cone_pdf(local_dir, (Float)m_cos_cutoff_angle);
+
+        // 2. Sample spectrum
+        auto [wavelengths, weights] = sample_wavelength<Float, Spectrum>(wavelength_sample);
+        auto falloff_spec = falloff_curve(local_dir, wavelengths, active);
+
+        return { Ray3f(trafo * m_position, trafo * local_dir, time, wavelengths),
+                unpolarized<Spectrum>(falloff_spec / pdf_dir) };
+    }
+
+    std::pair<DirectionSample3f, Spectrum> sample_direction(const Interaction3f &it,
+                                                            const Point2f &/*sample*/,
+                                                            Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleDirection, active);
+
+        auto trafo = m_world_transform->eval(it.time, active);
+
+        DirectionSample3f ds;
+        ds.p        = trafo * m_position;
+        ds.n        = 0.f;
+        ds.uv       = 0.f;
+        ds.time     = it.time;
+        ds.delta    = true;
+        ds.object   = this;
+        ds.d        = ds.p - it.p;
+        ds.dist     = norm(ds.d);
+        Float inv_dist = rcp(ds.dist);
+        ds.d        *= inv_dist;
+        Vector3f local_d = trafo.inverse() * ds.d;
+        ds.pdf      = warp::square_to_uniform_cone_pdf(local_d, (Float)m_cos_cutoff_angle);
+        Spectrum falloff_spec = falloff_curve(local_d, it.wavelengths, active);
+
+        return { ds, unpolarized<Spectrum>(falloff_spec * (inv_dist * inv_dist) / ds.pdf) };
+    }
+
+    Float pdf_direction(const Interaction3f &, const DirectionSample3f &ds, Mask active) const override {
+        auto trafo = m_world_transform->eval(ds.time, active);
+        return warp::square_to_uniform_cone_pdf(trafo.inverse() * ds.d, 
+                                                (Float)m_cos_cutoff_angle);
+    }
+
+    Spectrum eval(const SurfaceInteraction3f &si, Mask active) const override { 
+        auto trafo = m_world_transform->eval(si.time, active);
+        Vector3f d = trafo * m_position - si.p;
+        Float inv_dist = rcp(norm(d));
+        d *= inv_dist;
+        Spectrum falloff_spec = falloff_curve(trafo.inverse() * d, si.wavelengths, active);
+
+        return unpolarized<Spectrum>(falloff_spec * inv_dist * inv_dist);
```

Comment:
I just realized I was wrong. Those two methods should indeed return `0.f`. I kept thinking about this implementation this other way (e.g. `sample_ray`). This should make things easier now ;)

Reply:
Okay, I will make `eval()` and `pdf_direction()` make return 0. But just out of curiosity - why do you think these should return 0 and why did you think they should not earlier? Just want to learn as I am making another kind of light and knowing this will help.

Reply:
First I was thinking of `pdf_direction()` as the probability of sampling a specific direction **from the light source**. (e.g. what `sample_ray()` does).

But this method actually compute the probability of sampling a direction **towards the light source** from a surface (e.g. what `sample_direction()` does).

As this light source is a `DeltaPositionEmitter` (just a point), it is impossible to hit it *by chance*. In other words, the probability of sampling a direction towards this emitter via other methods than calling `sample_direction()` on it is `0.f`.

Does this make more sense to you?

Reply:
Yes, crystal clear. Thanks a lot.